src/ui/public/state_management/state_storage/_tests__/hashed_item_store.js

```diff
@@ -0,0 +1,190 @@
+          hashedItemStore = new HashedItemStore(sessionStorage);
+
+          // Add some items that will be removed.
+          hashedItemStore.setItem('1', item);
+          setTimeout(() => {
+            // Do this a little later so that this item is newer.
+            hashedItemStore.setItem('2', item);
+            // Cap the storage at its current size.
+            storageSizeLimit = sessionStorage.getStubbedSize();
+            sessionStorage.setStubbedSizeLimit(storageSizeLimit);
+            done();
+          }, 1);
+        });
+
+        describe('and the item will fit', () => {
+          it('removes older items until the new item fits', () => {
+            hashedItemStore.setItem(hash, item);
+            expect(sessionStorage.getItem('1')).to.equal(null);
+            expect(sessionStorage.getItem('2')).to.equal(item);
+          });
+
+          it('persists the item in sessionStorage', () => {
+            hashedItemStore.setItem(hash, item);
+            expect(sessionStorage.getItem(hash)).to.equal(item);
+          });
+
+          it('returns true', () => {
+            const result = hashedItemStore.setItem(hash, item);
+            expect(result).to.equal(true);
+          });
+        });
+
+        describe(`and the item won't fit`, () => {
+          let itemTooBigToFit;
+
+          beforeEach(() => {
+            // Make sure the item is longer than the storage size limit.
+            itemTooBigToFit = '';
+            const length = storageSizeLimit + 1;
+            for (let i = 0; i < length; i++) {
+              itemTooBigToFit += 'a';
+            }
+          });
+
+          it('removes all items', () => {
+            hashedItemStore.setItem(hash, itemTooBigToFit);
+            expect(sessionStorage.getItem('1')).to.equal(null);
+            expect(sessionStorage.getItem('2')).to.equal(null);
+          });
+
+          it(`doesn't persist the item in sessionStorage`, () => {
+            hashedItemStore.setItem(hash, itemTooBigToFit);
+            expect(sessionStorage.getItem(hash)).to.equal(null);
+          });
+
+          it('returns false', () => {
+            const result = hashedItemStore.setItem(hash, itemTooBigToFit);
+            expect(result).to.equal(false);
+          });
+        });
+      });
+    });
+  });
+
+  describe('#getItem', () => {
+    describe('if the item exists in sessionStorage', () => {
+      let sessionStorage;
+      let hashedItemStore;
+      const hash = 'a';
+      const item = JSON.stringify({});
+
+      beforeEach(() => {
+        sessionStorage = new StubBrowserStorage();
+        hashedItemStore = new HashedItemStore(sessionStorage);
+        hashedItemStore.setItem(hash, item);
+      });
+
+      it('returns the item', () => {
+        const retrievedItem = hashedItemStore.getItem(hash);
+        expect(retrievedItem).to.be(item);
+      });
+
+      it('removes the item last when freeing up storage space', async () => {
```

Comment:
I think the text here should probably read more like "marks the hash as touched so that it will be removed last when freeing up storage space". I also think that it's important to have more than 2 items in the store to show that happening.